libraries/AP_SmartAudio/AP_SmartAudio.h

```diff
@@ -0,0 +1,205 @@
+
+    // time the last_request is process
+    uint32_t last_request_sended_at=0;
+    // loops is waiting a response after a request
+    bool is_waiting_response=false;
+
+    // sends a frame over the wire
+    void send_request(smartaudioFrame_t requestFrame,uint8_t size);
+
+    // receives a frame response over the wire
+    void read_response(uint8_t *response_buffer,uint8_t inline_buffer_length);
+
+    // parses the response and updates the AP_VTX readings
+    bool parse_frame_response(const uint8_t *buffer);
+
+    // to prevent use of semaphores use an vtx_states_queue with 2 elements, so one buffered reading
+    ObjectBuffer<smartaudioSettings_s> vtx_states_queue{2};
+
+    // get last reading from the fifo queue
+    bool get_readings(AP_VideoTX *vtx_dest);
+
+    // enqueue a get settings request
+    void request_settings();
+
+    void set_operation_mode(uint8_t mode);
+
+    // enqueue a set frecuency request, specifiying if the setted frequency is for pit mode or not
+    void set_frequency(uint16_t frecuency,bool isPitModeFreq);
+
+    // enqueue a set channel request
+    void set_channel(uint8_t chan);
+
+    // enqueue a get pit mode frequency request
+    void get_pit_mode_frequency();
+
+    // enqueue a set power request using dbm
+    void set_power_dbm(uint8_t power);
+
+    void set_power_mw(uint16_t power_mw);
+
+    /**
+     *
+     * */
+    void print_bytes_to_hex_string(uint8_t buf[],uint8_t x);
+    // {
+    //     int i;
+    //     for (i = 0; i < x; i++) {
+    //         if (i > 0) {
+    //             printf(":");
+    //         }
+    //         printf("%02X", buf[i]);
+    //     }
+    //     printf("\n");
+    // }
+
+       // utility method to get power in dbm unit from the settled power level
+    static uint8_t _get_power_in_dbm_from_vtx_power_level(uint8_t power_level, uint8_t& protocol_version, uint8_t& power_in_dbm);
+
+        // utility method to get power in dbm unit from the settled power level
+    static uint8_t _get_power_in_dbm_from_vtx_power_level(uint8_t power_level, uint8_t& protocol_version);
+
+    // utility method to get power level which corresponds to a dbm power defined
+    static uint8_t _get_power_level_from_dbm(uint8_t sma_version,uint8_t power);
+
+private:
+    // serial interface
+    AP_HAL::UARTDriver *_port;                  // UART used to send data to SmartAudio VTX
+
+    //Pointer to singleton
+    static AP_SmartAudio* singleton;
+
+    // get current state, first on fifo queue
+    smartaudioSettings_t* _get_current_state(smartaudioSettings_t *stateStorage){
+         
+         
+         if(!vtx_states_queue.is_empty()){
+            
+            vtx_states_queue.peek(stateStorage,1);
+         }else{
+             stateStorage=nullptr;
+         }
+         
+        return stateStorage;
+    }
+
+    void _print_state(smartaudioSettings_t *state);
+
+
+
+static uint16_t _get_power_in_mw_from_dbm(uint8_t power){
+     switch (power) {
```

Comment:
Too big to be inline

Reply:
Perhaps we could turn these into a table in AP_VideoTX, I'm sure there will be more

Reply:
Not already in AP_VideoTX, waiting your revision prior to move. ;)